gsy_framework/sim_results/electric_blue/volume_timeseries.py

```diff
@@ -100,44 +120,41 @@ def _add_total_energy_sold(
+        average_sell_price = asset_stats.average_sell_price
+        for time_slot, energy in profile.items():
+            self._add_to_volume_time_series(
+                time_slot,
+                {
+                    "energy_kWh": energy,
+                    "price": energy * average_sell_price
+                }, product_type, "sold")
 
     def _add_to_volume_time_series(
-            self, time_slot: DateTime, value: float,
+            self, time_slot: DateTime, time_slot_info: Dict,
             product_type: AvailableMarketTypes, attribute_name: str):
         """Add time slot value to the correct time slot to asset volume time series."""
         time_slot = self._adapt_time_slot(time_slot)
         year = time_slot.start_of("year")
-        volume_time_series = self._get_asset_volume_time_series(year=year)
-        volume_time_series[str(time_slot)][product_type.name][attribute_name] += value
-        self._asset_volume_time_series_buffer[year] = volume_time_series
-
-    def _adapt_time_slot(self, time_slot: DateTime) -> DateTime:
-        """Find the containing time slot of the smaller time slot."""
-        if self.resolution == AggregationResolution.RES_15_MINUTES:
-            return time_slot.set(minute=(time_slot.minute // 15) * 15)
-        return time_slot.start_of(START_OF[self.resolution])
-
-    def _get_asset_volume_time_series(self, year: DateTime) -> Dict[str, Dict]:
-        """Return asset volume time series for the required year.
-        If not found in the buffer, it tries fetching it from DB.
-        If not found in the DB, it will generate a new one for the whole year."""
-        if year in self._asset_volume_time_series_buffer:
-            time_series = self._asset_volume_time_series_buffer[year]
-        else:
-            time_series = self._fetch_asset_volume_time_series_from_db(year)
-            if time_series is None:
-                time_series = {
-                    str(ts): self._get_time_series_template(value)
-                    for ts, value in self._generate_SSP_time_series(year)}
-        return time_series
-
-    def _fetch_asset_volume_time_series_from_db(
-            self, year: DateTime) -> Optional[Dict[str, Dict]]:
-        """Fetch already saved asset volume time series."""
-        # TODO: should be implemented.
-        return None
+        volume_time_series = self._get_asset_time_series(year=year)
+
+        # UPDATING energy_kWh, price and energy_rate
+        attribute_data = volume_time_series[str(time_slot)][product_type.name][attribute_name]
+        attribute_data["energy_kWh"] = round_floats_for_ui(
+            attribute_data["energy_kWh"] + time_slot_info["energy_kWh"])
+        attribute_data["price"] = round_prices_to_cents(
+            attribute_data["price"] + time_slot_info["price"])
+        try:
+            attribute_data["energy_rate"] = round_prices_to_cents(
+                attribute_data["price"] / attribute_data["energy_kWh"])
```

Comment:
I guess it is a requirement to return both "price" and "energy_rate", right? AFAIU, either of them can easily be calculated by the FE and we can save the DB space.

Reply:
In order to calculate the `energy_rate`, I needed it's compnents (`energy_kWh` and `price`) to be present, otherwise we can't update the `energy_rate` every 15 minutes. And you're correct about saving space, FE can easilly calculate `energy_rate` using price and energy_kWh. The reason I did it in backend was to not letting FE to enter our business logic. What do you think?

Reply:
> The reason I did it in backend was to not letting FE to enter our business logic

IMO they can to a simple devision without putting too much effort on the FE. Maybe @spyrostz begs to differ?

Reply:
Well, I am with Amirhossein on this one. I know that it is not a big deal for the UI to do a simple division, however on timeseries data these are going to be *some* divisions. The calculation time of these will essentially be passed on to the user, and judging by the performance of the UI, I think that we should do our best to decrease as much as possible the computations that the UI has to perform.

Reply:
OK